lib/afc/index.js

```diff
@@ -0,0 +1,237 @@
+/* eslint-disable promise/prefer-await-to-callbacks */
+import LengthBasedSplitter from '../util/transformer/length-based-splitter';
+import { Operations, operationCode, Errors, errorCode } from './protocol';
+import AfcEncoder from './transformer/afcencoder';
+import AfcDecoder from './transformer/afcdecoder';
+import B from 'bluebird';
+import _ from 'lodash';
+
+const AFC_SERVICE_NAME = 'com.apple.afc';
+
+const NULL_DELIMETER_CODE = 0x00;
+
+const KB = 1024;
+const MB = KB * KB;
+
+class AfcService {
+
+  constructor (socketClient) {
+    this._socketClient = socketClient;
+    this._splitter = new LengthBasedSplitter(true, 10 * MB, 8, 8, -8);
+    this._decoder = new AfcDecoder();
+    this._socketClient.pipe(this._splitter).pipe(this._decoder);
+
+    this._encoder = new AfcEncoder();
+    this._encoder.pipe(this._socketClient);
+
+    this._responseCallbacks = {};
+
+    this._packetNumber = 0;
+    this._decoder.on('data', this.handleData.bind(this));
+  }
+
+  handleData (data) {
+    const cb = this._responseCallbacks[data.packetNumber];
+    if (!_.isFunction(cb)) {
+      return;
+    }
+    cb(data);
+  }
+
+  async createDirectory (path) {
+    const {packetNumber, response} = this.createPacketPromise();
+
+    const data = {
+      opCode: Operations.MAKE_DIR,
+      packetNumber,
+      headerPayload: Buffer.from(path)
+    };
+    this._encoder.write(data);
+    const res = await response;
+    this.checkStatus(res);
+  }
+
+  async deleteDirectory (path) {
+    const {packetNumber, response} = this.createPacketPromise();
+
+    const data = {
+      opCode: Operations.REMOVE_PATH_AND_CONTENTS,
+      packetNumber,
+      headerPayload: Buffer.from(path)
+    };
+    this._encoder.write(data);
+    const res = await response;
+    this.checkStatus(res);
+  }
+
+  async listDirectory (path) {
+    const {packetNumber, response} = this.createPacketPromise();
+
+    const data = {
```

Comment:
this blocks looks very similar in all these methods. Would it be possible to somehow unify it?

Reply:
Will see what I can do